packages/expo-sms/ios/ExpoSMSModule.swift

```diff
@@ -0,0 +1,97 @@
+import ExpoModulesCore
+import MessageUI
+import CoreServices
+import MobileCoreServices
+import UniformTypeIdentifiers
+
+struct ExpoSMSContext {
+  let promise: Promise
+  let smsDelegate: SMSDelegate
+}
+
+public class ExpoSMSModule: Module, SMSResultHandler {
+  private var smsContext: ExpoSMSContext?
+  private lazy var utils = appContext?.utilities
+
+  public func definition() -> ModuleDefinition {
+    Name("ExpoSMS")
+
+    AsyncFunction("isAvailableAsync") {
+      return MFMessageComposeViewController.canSendText()
+    }
+
+    AsyncFunction("sendSMSAsync") {(addresses: [String], message: String, options: SmsOptions, promise: Promise) in
+      try sendSMSAsync(addresses: addresses, message: message, options: options, promise: promise)
+    }.runOnQueue(.main)
+  }
+
+  private func sendSMSAsync(addresses: [String], message: String, options: SmsOptions, promise: Promise) throws {
+    if !MFMessageComposeViewController.canSendText() {
+      throw SmsUnavailableException()
+    }
+
+    if smsContext != nil {
+      throw SmsPendingException()
+    }
+
+    let smsDelegate = SMSDelegate(handler: self)
+    let context = ExpoSMSContext(promise: promise, smsDelegate: smsDelegate)
+
+    let messageComposeViewController = MFMessageComposeViewController()
+    messageComposeViewController.messageComposeDelegate = context.smsDelegate
+    messageComposeViewController.recipients = addresses
+    messageComposeViewController.body = message
+
+    for attachment in options.attachments {
+      let utiRef = UTTypeCreatePreferredIdentifierForTag(
+        kUTTagClassMIMEType, attachment.mimeType as CFString, nil)
+
+      if utiRef == nil {
+        context.promise.reject("E_SMS_ATTACHMENT",
```

Comment:
To be honest I would recommend throwing an exception also here. In that cause you would need to extend `GenericException`.
Why exceptions are better? Because they're chained just like in Java — this would be the root cause, but when it's caught, it rethrows a `FunctionCallException` with a reference to the exception that caused it. It gives us and the user more context and clarifies what is the "throwing path". Also, they let us separate the logic from the error codes and messages.
One more note: promises can be rejected with an exception too!